### PR TITLE
Switch repo to a src layout and to Poetry

### LINKED ISSUE #40: Set up Python tooling
Set up standard Python tooling to simplify dependency management and developer onboarding. 

## Tooling - Virtual environments, Setuptools, and requirements.txt files

The current best practice for Python is for developers to develop in virtual environments (`venv`'s) and to capture project dependencies, entrypoints, and build/tool settings in a `pyproject.toml` file. Previously this meant understanding and using pip, setuptools, and a `requirements.txt` or `setup.py`/`setup.cfg` file.

While it is still possible to use setuptools & friends, (and pip is certainly still used) tooling has evolved and there are now popular tools like [Poetry](https://python-poetry.org/) available that take care of virtual environment creation, dependency management, and builds for you in a way similar to Node's NPM with its `package.json` and `package.lock` files.

Along with being able to quickly create isolated and repeatable dev environments tools like venv/setuptools and Poetry can also package your code for you into a wheel. Wheel's are advantageous as they make installing your code as simple as getting the `.whl` file on your target machine and running `pip install <file>.whl` and all dependencies will be set up for you.

There are also other non-python-specific package managers like anaconda/miniconda available that we may want to take advantage of on HPC.

## Code Organization

A number of prominent Python projects have also reorganized their source repos to contain all their application code under a `src/` directory. The reason they've done this is it makes it more difficult to unintentionally mess up include paths for modules and easier to set up and run the above tooling. (https://hynek.me/articles/testing-packaging/) I suspect having a layout like this would help #39. 

I'd propose a structure like the below where all of our modules live in a `src/` directory, tests are separated into a `tests` directory and shell scripts and docs go into their own top-level directories:

```
.
├── docs
├── pyproject.toml
├── scripts
│   └── shell_scripts.sh
├── src
│   ├── module1
│   ├── module2
│   └── moduleN
└── tests
    ├── module1
    ├── module2
    └── moduleN
```

## Do we want to do a spike to explore this?

I've set up my [VxBdpDemo project](https://github.com/NOAA-GSL/VxBdpDemo) to utilize Poetry and have been happy with it - would it be useful for me to do a few-day spike to investigate how it would work for this project? I do realize miniconda is heavily used on the HPC side and I believe there are ways to use poetry with conda.

### PR DESCRIPTION
**Pull request by ian-noaa**
This change switches the codebase to a `src` layout and switches us to use Poetry for our dependency management.

Major changes include:
* Switching to Python's [recommend `src` layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)
* Switching to [Poetry](https://python-poetry.org/) for dependency management
* Updating imports throughout the codebase to  work with the new structure
* Applying the [ruff](https://docs.astral.sh/ruff/) linter's safe fixes
* Updating & consolidating READMEs to keep them relevant

Things I'm planning to address in other PRs:
* Container updates
* Code formatting
* GitHub Actions

We can now do the following locally to setup the Python environment, run the test suite and run the ingest:

```bash
# Setup venv and install deps
poetry install

# Run tests
# Set CREDENTIALS to wherever your credentials file is
CREDENTIALS=config.yaml poetry run pytest tests/

# Run ingest
poetry run ingest # with appropriate CLI flags
```

Closes #40.

### CODE REVIEW

**Comment by ian-noaa** [comment]
@JeffHamiltonNOAA & @randytpierce - if you both could review this, I'd appreciate it. I've updated the PR description with highlights of the changes. For reference on how to get started with this repo and Poetry, I'd recommend referencing the [README.md ](https://github.com/NOAA-GSL/VxIngest/blob/switch-repo-to-src-layout/README.md) and the [Developer Guide](https://github.com/NOAA-GSL/VxIngest/blob/switch-repo-to-src-layout/docs/development-guide.md).

Let me know if you have any questions or want to pair to test this out!

**Comment by ian-noaa** [comment]
A note - This PR is pretty massive and for that, I apologize. For PRs like this, I personally find the "commit" tab useful for figuring out what was changed and why. I tried to keep the massive file moves in single commits and separate from code changes.

The test suite should run - I have instructions in [`tests/vxingest/README.md`](https://github.com/NOAA-GSL/VxIngest/blob/switch-repo-to-src-layout/tests/vxingest/README.md#running-tests) if you want to try.

**Review by randytpierce — COMMENTED** [review]
I don't see anything wrong structurally but we need to test this. In general I approve but want to test it before we merge. I need to build onmy local machine on this branch and run all the tests.

File: tests/vxingest/ctc_to_cb/test_int_metar_ctc.py [hunk 1]
```diff
@@ -126,7 +124,7 @@ def test_get_stations_geo_search():
         )
         ingest_document = ingest_document_result.content_as[dict]
         # instantiate a ctcBuilder so we can use its get_station methods
-        builder_class = getattr(ctc_builder, "CTCModelObsBuilderV01")
+        builder_class = ctc_builder.CTCModelObsBuilderV01
```
[the remark this answers is not in the document]
Comment (ian-noaa) [reply tests/vxingest/ctc_to_cb/test_int_metar_ctc.py:127]: Yeah, I thought it was interesting. It was picked up by the linter and it's one of the common "flake8-bugbear" rules. (B009) They had a decent page with an overview here: https://docs.astral.sh/ruff/rules/get-attr-with-constant/

**Comment by randytpierce** [comment]
I would like to run it locally as a script and as a container that I build
locally.
Randy

**Comment by ian-noaa** [comment]
Sounds good. I'd expect that the container won't work yet - I'm planning on addressing that in a separate PR. (And working on it now, actually) The local script should run via Poetry.

**Comment by ian-noaa** [comment]
Rebase on main to get Randy's latest changes. Once @randytpierce is okay with the test suite, (see #272) I'll go ahead and merge both PRs.

**Pull request merged**